core/src/main/java/com/yandex/yoctodb/util/immutable/impl/FoldedByteArrayIndex.java

```diff
@@ -0,0 +1,117 @@
+package com.yandex.yoctodb.util.immutable.impl;
+
+import com.yandex.yoctodb.util.buf.Buffer;
+import com.yandex.yoctodb.util.immutable.ByteArrayIndexedList;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.function.Function;
+
+public class FoldedByteArrayIndex implements ByteArrayIndexedList {
+    private final int elementCount;
+    private final Function<Integer, Integer> getOffsetIndex;
+    @NotNull
+    private final Buffer elements;
+    @NotNull
+    private final Buffer offsets;
+    @NotNull
+    private final Buffer indexes;
+
+    @NotNull
+    public static ByteArrayIndexedList from(
+            @NotNull final Buffer buf) {
+        final int elementsCount = buf.getInt();
+        final int offsetsCount = buf.getInt();
+
+        int sizeOfIndexOffsetValue;
+        if (offsetsCount <= 255) { // one byte 2^8 - 1 = 127
+            sizeOfIndexOffsetValue = 1;
+        } else if (offsetsCount <= 65535) {  // to  2^16 - 1 = 65535
+            sizeOfIndexOffsetValue = 2;
+        } else {
+            sizeOfIndexOffsetValue = 4;
+        }
+
+        // indexes of offsets
+        final Buffer indexes = buf.slice((elementsCount) * sizeOfIndexOffsetValue);
+
+        long shift = indexes.remaining();
+
+        // then offsets of element value
+        final Buffer offsets = buf.slice()
+                .position(shift)
+                .slice((offsetsCount) << 3);
+
+        shift = shift + offsets.remaining();
+
+        // then elements value
+        final Buffer elements = buf.slice()
+                .position(shift)
+                .slice();
+
+        return new FoldedByteArrayIndex(
+                elementsCount,
+                offsetsCount,
+                offsets,
+                elements,
+                indexes);
+    }
+
+    private FoldedByteArrayIndex(
+            final int elementCount,
+            final int offsetsCount,
+            @NotNull final Buffer offsets,
+            @NotNull final Buffer elements,
+            @NotNull final Buffer indexes) {
+        assert elementCount >= 0 : "Negative element count";
+
+        this.elementCount = elementCount;
+        this.elements = elements;
+        this.offsets = offsets;
+        this.indexes = indexes;
+
+        if (offsetsCount <= 255) { // one byte 2^8 - 1 = 127
```

Comment:
Again magic number. There is so much benefits of extracting them into a constants :)